File: src/melange/resolve.ml
(* Map request paths to HTML / Markdown siblings and decide passthrough. *)

external decode_uri_component : string -> string = "decodeURIComponent"

let static_ext =
  [%mel.re
    "/\\.(?:css|js|mjs|map|png|jpe?g|webp|gif|svg|avif|ico|woff2?|ttf|otf|eot|xml|txt|json|pdf|mp4|webm|mp3|wav|ogg|zip|wasm)$/i"]

let rec rtrim_slashes s =
  let n = String.length s in
  if n > 0 && s.[n - 1] = '/' then rtrim_slashes (String.sub s 0 (n - 1))
  else s

let before_delim s delim =
  match String.index_opt s delim with
  | None -> s
  | Some i -> String.sub s 0 i

let normalize_path pathname =
  if pathname = "" then "/"
  else
    try
      let decoded = decode_uri_component pathname in
      let clean = decoded |> fun s -> before_delim s '?' |> fun s -> before_delim s '#' in
      if clean = "" || clean = "/" then "/"
      else
        let trimmed = rtrim_slashes clean in
        if trimmed = "" then "/" else trimmed
    with _ -> if pathname = "" then "/" else pathname

let last_segment clean =
  match List.rev (String.split_on_char '/' clean) with
  | [] -> ""
  | last :: _ -> last

(* Assets and already-negotiated files skip Accept handling. *)
let is_passthrough_path pathname =
  let clean = normalize_path pathname in
  if clean = "/" then false
  else
    let last = last_segment clean in
    if not (String.contains last '.') then false
    else if Js.Re.test ~str:last [%mel.re "/\\.html?$/i"] then false
    else Js.Re.test ~str:last static_ext || String.ends_with ~suffix:".md" last

let html_candidates pathname =
  let clean = normalize_path pathname in
  if clean = "/" then [ "/index.html" ]
  else [ clean ^ ".html"; clean ^ "/index.html" ]

let markdown_candidates pathname =
  let clean = normalize_path pathname in
  if clean = "/" then [ "/index.md" ]
  else [ clean ^ ".md"; clean ^ "/index.md" ]

let is_not_found_probe pathname =
  let clean = normalize_path pathname in
  clean = "/__missing_agentic_404__" || String.starts_with ~prefix:"/__missing" clean

File: src/negotiate/resolve.js
/**
 * Map request paths to HTML / Markdown siblings and decide passthrough.
 */

const STATIC_EXT =
  /\.(?:css|js|mjs|map|png|jpe?g|webp|gif|svg|avif|ico|woff2?|ttf|otf|eot|xml|txt|json|pdf|mp4|webm|mp3|wav|ogg|zip|wasm)$/i;

/**
 * @param {string} pathname
 */
export function normalizePath(pathname) {
  if (!pathname) return "/";
  try {
    const decoded = decodeURIComponent(pathname);
    const clean = decoded.split("?")[0].split("#")[0];
    if (clean === "" || clean === "/") return "/";
    return clean.replace(/\/+$/, "") || "/";
  } catch {
    return pathname || "/";
  }
}

/**
 * Assets and already-negotiated files skip Accept handling.
 * @param {string} pathname
 */
export function isPassthroughPath(pathname) {
  const clean = normalizePath(pathname);
  if (clean === "/") return false;
  const last = clean.split("/").pop() || "";
  if (!last.includes(".")) return false;
  if (/\.html?$/i.test(last)) return false;
  return STATIC_EXT.test(last) || last.endsWith(".md");
}

/**
 * Candidate HTML files for a URL (first existing wins).
 * @param {string} pathname
 * @returns {string[]}
 */
export function htmlCandidates(pathname) {
  const clean = normalizePath(pathname);
  if (clean === "/") return ["/index.html"];
  return [`${clean}.html`, `${clean}/index.html`];
}

/**
 * Candidate Markdown siblings for a URL.
 * @param {string} pathname
 * @returns {string[]}
 */
export function markdownCandidates(pathname) {
  const clean = normalizePath(pathname);
  if (clean === "/") return ["/index.md"];
  return [`${clean}.md`, `${clean}/index.md`];
}

/**
 * @param {string} pathname
 */
export function isNotFoundProbe(pathname) {
  const clean = normalizePath(pathname);
  return clean === "/__missing_agentic_404__" || clean.startsWith("/__missing");
}
